In [4]:

#Importing required library
#import pandas as pd and Reading a CSV file into a DataFrame

import pandas as pd

df_1k = pd.read_csv("/content/synthetic_diabetes_1000.csv")
df_10k = pd.read_csv("/content/synthetic_diabetes_10000.csv")
df_100k = pd.read_csv("/content/synthetic_diabetes_1e+05.csv")


In [12]:
import pandas as pd
df = pd.read_csv("/content/synthetic_diabetes_1000.csv")
print(df.columns)


Index(['pregnant', 'glucose', 'pressure', 'triceps', 'insulin', 'mass',
       'pedigree', 'age', 'outcome'],
      dtype='object')


In [ ]:
#  1000

#  1 hidden layer 4 nodes

In [13]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_1000.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 1 hidden layer with 4 nodes
model = Sequential()
model.add(Dense(4, input_dim=X.shape[1], activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.0875
Validation Error: 0.0850
Time of Execution: 4.51 seconds


In [ ]:
#10000

#1 hidden layer 4 nodes

In [14]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_10000.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 1 hidden layer with 4 nodes
model = Sequential()
model.add(Dense(4, input_shape=(X.shape[1],), activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.0035
Validation Error: 0.0060
Time of Execution: 15.19 seconds


In [15]:
#100000

#1 hidden layer 4 nodes

In [16]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_1e+05.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 1 hidden layer with 4 nodes
model = Sequential()
model.add(Dense(4, input_shape=(X.shape[1],), activation='relu'))
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.0012
Validation Error: 0.0013
Time of Execution: 150.12 seconds


In [17]:
#1000	2 hidden layers, 4 nodes each

In [18]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_1000.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 2 hidden layers, 4 nodes each
model = Sequential()
model.add(Dense(4, input_shape=(X.shape[1],), activation='relu'))
model.add(Dense(4, activation='relu'))  # Second hidden layer
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.1025
Validation Error: 0.0700
Time of Execution: 5.93 seconds


In [19]:
#10000	2 hidden layers, 4 nodes each

In [20]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_10000.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 2 hidden layers, 4 nodes each
model = Sequential()
model.add(Dense(4, input_shape=(X.shape[1],), activation='relu'))
model.add(Dense(4, activation='relu'))  # Second hidden layer
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.0038
Validation Error: 0.0070
Time of Execution: 17.02 seconds


In [21]:
#100000	2 hidden layers, 4 nodes each

In [22]:
# Load dataset
df = pd.read_csv("synthetic_diabetes_1e+05.csv")
X = df.drop("outcome", axis=1)
y = df["outcome"]

# Train/test split and scaling
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)

# Build model: 2 hidden layers, 4 nodes each
model = Sequential()
model.add(Dense(4, input_shape=(X.shape[1],), activation='relu'))
model.add(Dense(4, activation='relu'))  # Second hidden layer
model.add(Dense(1, activation='sigmoid'))

model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train model and track time
start = time.time()
history = model.fit(X_train, y_train, validation_data=(X_val, y_val), epochs=20, verbose=0)
end = time.time()

# Results
training_error = 1 - history.history['accuracy'][-1]
validation_error = 1 - history.history['val_accuracy'][-1]
execution_time = round(end - start, 2)

print(f"Training Error: {training_error:.4f}")
print(f"Validation Error: {validation_error:.4f}")
print(f"Time of Execution: {execution_time} seconds")


/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Training Error: 0.0017
Validation Error: 0.0017
Time of Execution: 162.76 seconds


In [23]:
# 2. Based on the results, which model do you consider as superior, among the deep learning models fit?



 The evaluation process involved testing six model configurations against three dataset sizes (1,000;  10,000; and 100,000 rows) and two architectures (1 hidden layer vs.  2 hidden layers, each with 4 nodes).

The results showed that the model with 1 hidden  layer and 4 nodes achieved the lowest training and validation errors (0.0012 and  0.0013 respectively) when trained on the 100,000-row dataset. This model trained faster  than the equivalent 2-layer model at the same data size which needed 162.76 seconds to  train compared to 150.12 seconds.

The 2-hidden-layer configurations achieved comparable accuracy levels but  failed to deliver meaningful improvements in validation performance to warrant increased complexity and longer training times. The  2-layer model trained on 100,000 rows produced a validation error of 0.0017 and  needed more time to complete.

The models demonstrated higher validation errors and unstable learning behavior and overfitting  at smaller data sizes (e.g., 1,000 rows) which shows that deep learning models  need larger datasets to perform well.

The deep learning model that uses 1 hidden layer with 4  nodes and trains on 100,000 rows of data stands as the best option. The model achieves  the highest accuracy with the lowest validation error and shortest training time among the tested configurations.